File: frontend/fetch_21st.py
import urllib.request
from bs4 import BeautifulSoup
import json
import ssl
import re

def fetch_component(url):
    ctx = ssl.create_default_context()
    ctx.check_hostname = False
    ctx.verify_mode = ssl.CERT_NONE
    
    req = urllib.request.Request(url, headers={'User-Agent': 'Mozilla/5.0'})
    html = urllib.request.urlopen(req, context=ctx).read()
    soup = BeautifulSoup(html, 'html.parser')
    
    scripts = soup.find_all('script')
    for script in scripts:
        if script.string and 'about-page' in script.string and 'code.1759946300968.tsx' in script.string:
            # We found the script tag with the payload. Let's find the actual code string.
            # The code is usually assigned to a variable or part of a JSON-like structure.
            # Look for the "code":"..." pattern or similar.
            try:
                # Naive regex extraction of the code block. 21st.dev stores it as a literal string in the array.
                match = re.search(r'"code":"(.*?)"(?:,"demoCode"|,"dependencies")', script.string)
                if match:
                    codeRaw = match.group(1)
                    # Decode unicode escapes and newlines
                    code = codeRaw.replace('\\n', '\n').replace('\\"', '"').replace('\\\\', '\\')
                    
                    with open('src/components/AboutPage.tsx', 'w') as f:
                        f.write(code)
                    print("Successfully extracted code to src/components/AboutPage.tsx")
                    return
            except Exception as e:
                print(f"Error parsing: {e}")
                
    print("Could not find code block in scripts.")

if __name__ == '__main__':
    fetch_component('https://21st.dev/ruixenui/about-page/default')

File: frontend/fetch_21st.js
const https = require('https');
const fs = require('fs');

function fetchCode(url, filename) {
    https.get(url, { rejectUnauthorized: false }, (res) => {
        let data = '';
        res.on('data', chunk => data += chunk);
        res.on('end', () => {
            // Find __NEXT_DATA__
            let match = data.match(/<script id="__NEXT_DATA__" type="application\/json">(.*?)<\/script>/);
            if (match) {
                let j = JSON.parse(match[1]);
                // traverse j to find "code" key
                function findCode(obj) {
                    if (!obj || typeof obj !== 'object') return null;
                    if (obj.code && typeof obj.code === 'string' && obj.code.includes('export ')) return obj.code;
                    for (let k in obj) {
                        let res = findCode(obj[k]);
                        if (res) return res;
                    }
                    return null;
                }
                let code = findCode(j);
                if (code) {
                    fs.writeFileSync('src/components/ui/' + filename, code);
                    console.log('Saved', filename);
                } else {
                    console.log('Code not found in JSON for', url);
                }
            } else {
                console.log('__NEXT_DATA__ not found for', url);
            }
        });
    });
}

fetchCode('https://21st.dev/osmosupply/parallax-scrolling/default', 'parallax-scrolling.tsx');
